Fix sub-group broadcast test

AlexeySachkov opened the pull request:
The test made incorrect assumptions about work-items mapping to sub-groups, i.e. it expected that sub-groups are formed from subsequent work-items and thus the last element in the first sub-group of an nd-range would have its global linear id equal to sub-group size.

However, SYCL specification does not mandate any specific mapping of work-items into sub-groups and therefore assumptions made by test are incorrect.

This PR rewrites the test so that kernel stores both original and broadcasted values into corresponding buffers and the former is used as a reference for the latter and checked on host.

keryell commented on tests/group_functions/group_broadcast.h:
```diff
@@ -131,87 +131,127 @@ void broadcast_sub_group(sycl::queue& queue) {
-
-          // broadcast from the first workitem
-          local_var = splat_init<T>(item.get_global_linear_id() + 1);
+          // Find local id of first, last and some third sub-group item in
```
```suggestion
          // Find local id of first, last and some third sub-group id in
```
To be less confusing, since `sycl::item` is also a thing.
I had to look at the spec to refresh my memory on sub-groups.
It might be clearer to add a comment before `first_id` to say it is the sub-group leader or `sub_group::leader()`?

AlexeySachkov replied:
Reworded the comment in 2952e4e0fee35fe8d3d4ff9a23b5e48abd4ff311

keryell commented on tests/group_functions/group_broadcast.h:
```diff
@@ -131,87 +131,127 @@ void broadcast_sub_group(sycl::queue& queue) {
+          // broadcasting and receiving work-items
+          sycl::id<1> first_id = 0;
+          sycl::id<1> mid_id = sub_group.get_local_range() / 2;
+          sycl::id<1> last_id = sub_group.get_local_range();
+          --last_id[0];
```
Just curious about this in 2 lines instead of just 1?

AlexeySachkov replied:
I just didn't think about changing the existing code much. Shortened to a single line in 2952e4e0fee35fe8d3d4ff9a23b5e48abd4ff311

keryell commented on tests/group_functions/group_broadcast.h:
```diff
@@ -131,87 +131,127 @@ void broadcast_sub_group(sycl::queue& queue) {
+
+          broadcasted_value =
+              sycl::group_broadcast(sub_group, value_to_broadcast,
+                                    sub_group.get_local_linear_range() - 1);
```
Is it not possible to use `last_id` here?

AlexeySachkov replied:
I should probably leave a comment here, because `group_broadcast` has overloads accepting both `size_t` and `sycl::it` as an index of a work-item which should be used as source.

Were are trying to test both interfaces here. I should also fix the last argument in `ASSERT_RETURN_TYPE` to match that particular overload with plain `size_t`

keryell replied:
The more comments I read, the happier my brain is. ;-)

AlexeySachkov replied:
Applied all the mentioned changes in 2952e4e0fee35fe8d3d4ff9a23b5e48abd4ff311

keryell reviewed (commented):
Thanks for cleaning this recurring `sub_group`-related bug.

AlexeySachkov commented:
2952e4e0fee35fe8d3d4ff9a23b5e48abd4ff311 addresses all comments, but I haven't really tried to run the test using that commit, because my machine misbehaves. That new revision seems to be almost NFC, so I don't expect that it introduces any regressions.

Note that I won't be available to react to feedback next week, but maybe @steffenlarsen will back me up in case of further comments/questions.

keryell commented on tests/group_functions/group_broadcast.h:
```diff
@@ -164,13 +164,12 @@ void broadcast_sub_group(sycl::queue& queue) {
         if (sub_group.get_group_id()[0] == 0) {
-          // Find local id of first, last and some third sub-group item in
-          // between. Will be used to check different combinations of
+          // Find local id of a leader, last and some other work-item in the
```
```suggestion
          // Find local id of the leader, last and some other work-item in the
```
"the leader" in the sense of the `.leader()`?

keryell reviewed (commented):
Great! Quite clearer.

keryell reviewed (approved):
Thanks!
